### DISCUSSION TITLE
Guidance regarding on-disk compatibility with inet extension IPv6 update

### DISCUSSION THREAD

**Discussion by troycurtisjr** [duckdb/duckdb#10581] [General]
I have a [functional ipv6 implementation](https://github.com/troycurtisjr/duckdb/tree/add-ipv6-inet), but to do it the most natural way I ended up changing the "address" field of the struct which comprises the `inet` logical type from `hugeint_t` to `uhugeint_t`. This matches the type in the `IPAddress` utility class, and certainly makes the most sense as the underlying IPv6 type. I was able to get the full implementation with tests working great with this change.

The trouble comes into play when I load a DB file from a duckdb version from before this change. There is no error, but it ends up treating it just like a `struct(uint8_t, hugeint_t, uint16_t)` type, though it labels it with the `inet` alias. However, the various new overloads don't get executed, presumably because they are registered now with `struct(uint8_t, uhugeint_t, uint16_t)`.

I'm trying to determine the best way to deal with this situation, and I would like to know the preferences of the team.

One option is to not be backwards-compatible, requiring users to dump and load any existing db files similar to what would need to be done if the file format version changes. The biggest downside here is that it isn't super obvious that this needs to happen. Since the file will load fine, and the type of the column is even labeled as "inet", yet it actually doesn't get treated like an inet:

```
D .schema
CREATE TABLE inettest(id INTEGER, addr INET);
D select * from inettest;
┌───────┬───────────────────────────────────────────────────┐
│  id   │                       addr                        │
│ int32 │                       inet                        │
├───────┼───────────────────────────────────────────────────┤
│     1 │ {'ip_type': 1, 'address': 2130706433, 'mask': 32} │
│     2 │                                                   │
│     3 │ {'ip_type': 1, 'address': 4294967295, 'mask': 31} │
│     4 │ {'ip_type': 1, 'address': 0, 'mask': 0}           │
│     5 │ {'ip_type': 1, 'address': 3221225984, 'mask': 24} │
└───────┴───────────────────────────────────────────────────┘
D select id,addr::inet from inettest;
┌───────┬────────────────────┐
│  id   │ CAST(addr AS inet) │
│ int32 │        inet        │
├───────┼────────────────────┤
│     1 │ 127.0.0.1          │
│     2 │                    │
│     3 │ 255.255.255.255/31 │
│     4 │ 0.0.0.0/0          │
│     5 │ 192.0.2.0/24       │
└───────┴────────────────────┘
D insert into inettest values (7, '::1'), (8, '2345:425::5673:23b5/64');
Error: Conversion Error: Type VARCHAR with value '::1' can't be cast to the destination type STRUCT
```

I thought another option might be to just use the old on-disk format, since the bytes stored can be interpreted in a way which can work. I even got an [implementation mostly working with this approach](https://github.com/troycurtisjr/duckdb/tree/add-ipv6-inet-compat). Unfortunately, it doesn't work correctly when sorting, since the `address` value is treated as a signed value. I was digging around for a while trying to figure out how to overload comparison operations, but it seems that [you can't](https://github.com/duckdb/duckdb/discussions/7473) in extensions.

A third option I thought of was trying to implement compatible functions with the old struct type so that things will mostly work when loading the old data. I think for read and display purposes I might be able to get this to work as expected, however it would start to fall apart if the user then tried to take advantage of the new IPv6 support and write IPv6 values into the legacy column. Like the second option above, things would mostly appear to work, except that ordering would be wrong. Worse, a lot of ordering would be right, it would only ever be wrong once the top bit of the first word was set, then it would become nonsensical for IP addresses. This makes it really likely that someone won't immediately realize something is off and introduce subtle bugs in application's use.

I have also tried a few permutations of having a different compatible "type", but so far I haven't had any luck. Since the on-disk bytes between the `uhugeint_t` and `hugeint_t` would be the same, and when the `hugeint_t` is used there are only ipv4 addresses and thus no worry about ordering, I thought the same new functions could just work transparently for both, and it is only when trying to store a ipv6 into an old format that an error would be generated. However, so far I haven't figured out how to make that work. It seems like the only way of supporting both on-disk formats is to leave the `inet` type name mapped to the old struct, and create a new type for the new format. Something like maybe `inet6`, though then that makes it seem like it is ONLY for ipv6 so maybe `inet2` or `inet_new` or something like that. 

So I guess I have 2 main questions:

1. Should I continue trying to find a way to support the old and new formats at the same time, or just state that existing DB files will need to be dumped and reloaded in order to support ipv6.
2. Is there some mechanism I'm missing where I can specify 2 different struct type option for the same type? Something similar to the way overloads can have one name, but many type dependent functions?

Relevant discussion on ipv6 support in inet [in another post](https://github.com/duckdb/duckdb/discussions/2945#discussioncomment-6215857).

**Comment by Mytherin** [comment]
While not ideal, it should be possible to use `hugeint_t` for storing IPv6 addresses as it is "essentially" the same as `uhugeint_t` except it has a signed bit. While `uhugeint_t` is a better fit - we use `hugeint_t` to do this for `UUID` as well as we did not have `uhugeint_t` until recently. Is it not possible to add a few utility functions to convert IPv6 to/from `hugeint_t` to keep this compatible with the old storage format?

**Reply by troycurtisjr** [reply]
I know I posted a wall of text, so it was probably missed in all the verbiage, but I have the working version of storage as hugeint_t. It *mostly* works in that you can store and retrieve the values just fine. The trouble comes in with ordering, which orders by signed rules, which is nonsensical for IPV6 addresses. Namely `ffff::` would be before `7fff::` 😬

**Reply by Mytherin** [reply]
UUID faces the same issue - have a look at the code there. The solution is to flip the signed bit back and forth in the storage so that the `hugeint` ordering rules are consistent (i.e. you need to map `ffff::..` to the highest hugeint_t value, `7ffff:` to 0, and `0000::` to the lowest hugeint value).

**Reply by troycurtisjr** [reply]
oh interesting, I'll take a look tonight. I just knew that was some approach that I was missing, which is why I finally just posted a discussion topic to get some advice. Thanks!

**Reply by troycurtisjr** [reply]
It works perfectly! With a quick check to only flip the bit for IPv6 (so that the previously stored IPv4 values which won't have had the bit flip applied load correctly), we get good conversion, reasonable ordering, and backward compatibility! Thanks for the pointer. I'll get the code tidied up and a PR posted soon. Then I'll work on updating the docs to point out the new support in the other repo.

**Comment by uweber** [comment]
Is there a reason why inet_ntop/inet_pton is not used? Seems to work for me. See:
https://github.com/uweber/duckdb/commit/45688194d8d506f7cbe175ba521ad5a25b1a1fab

**Reply by troycurtisjr** [reply]
yes I had initially started down that pathway as well. My main concern was portability of the function, I know duckdb runs lots of places and I wasn't confident it was readily available everywhere, mainly concerned about things like Windows where I don't have much experience with. I didn't investigate too deeply since the existing existing had implemented it in-line, so I opted to follow in style. I generally prefer to use standard functions where available, so if there is no concern with portability that definitely seems like the way to go. I wouldn't see my hand-coding as a waste since it was fun to do 😁

**Reply by troycurtisjr** [reply]
So I did dig a bit deeper into this. As this function is POSIX it isn't in Windows, but there is an equivalent, with a different header and slightly modified API. Additionally, the returned values are in network byte order which may or may not be byte swapped relative to what a given architecture needs for correct integer operations. Lastly, there are a few additional details such as needing to ensure null terminations in the char arrays, which `string_t` doesn't seem to guarantee, which means copying to a new buffer, etc.

Considering the various portability concerns, which likely means not a lot of difference in lines of code after all the platform conditionals, along with the fact that we'd be using at least 3 different implementations of the underlying function, it actually seems like the in-line code is a better and more consistent choice.
